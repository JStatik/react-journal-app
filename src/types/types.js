const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start Loading',
    uiFinishLoading: '[UI] Finish Loading',

    notesNewEntry: '[Notes] New Note',
    notesActive: '[Notes] Set Active Note',
    notesAddNew: '[Notes] Add New Note',
    notesLoad: '[Notes] Load Notes',
    notesUpdate: '[Notes] Update Note',
    notesFileImg: '[Notes] Update Note Img',
    notesDelete: '[Notes] Delete Note',
    notesLogoutCleaning: '[Notes] Logout Cleaning'
}

export default types;
