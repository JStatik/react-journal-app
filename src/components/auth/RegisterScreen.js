import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithNameEmailPassword } from '../../actions/auth';
import useForm from '../../hooks/useForm';

const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector( store => store.ui );

    const [ formValues, handleInputChange, reset ] = useForm( { name: '', email: '', password: '', password2: '' } );
    const { name, email, password, password2 } = formValues;

    const isFormValid = () => {
        if( name.trim().length === 0 ) {
            dispatch( setError( 'Ingrese un nombre válido' ) );
            return false;
        } else if( !validator.isEmail( email ) ) {
            dispatch( setError( 'El email no es válido' ) );
            return false;
        } else if( password !== password2 ) {
            dispatch( setError( 'Las contraseñas no coinciden' ) );
            return false;
        } else if( password.length < 5 ) {
            dispatch( setError( 'La contraseña debe tener al menos 6 caracteres' ) );
            return false;
        }

        return true;
    };

    const handleRegister = ( event ) => {
        event.preventDefault();

        if( isFormValid() ) {
            dispatch( removeError() );
            dispatch( startRegisterWithNameEmailPassword( name, email, password ) );
        }

        reset();
    };

    return (
        <div className="animate__animated animate__fadeIn animate__faster">
            <h3 className="auth__title">Register</h3>

            <form onSubmit={ handleRegister }>
                { msgError && <div className="auth__alert-error text-center">{ msgError }</div> }

                <input type="text" name="name" value={ name } placeholder="Nombre" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="email" name="email" value={ email } placeholder="Email" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="password" name="password" value={ password } placeholder="Contraseña" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>
                <input type="password" name="password2" value={ password2 } placeholder="Repetir contraseña" autoComplete="off" className="auth__input" onChange={ handleInputChange }/>

                <button type="Submit" className="btn btn-primary btn-block mb-5" disabled={ loading }>Registrarse</button>

                <div className="text-center">
                    <Link to="/auth/login" className="link">Already registered?</Link>
                </div>
            </form>
        </div>
    )
};

export default RegisterScreen;
