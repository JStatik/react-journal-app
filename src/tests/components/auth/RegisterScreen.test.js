import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import RegisterScreen from '../../../components/auth/RegisterScreen';
import types from '../../../types/types';

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const store = mockStore( {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
} );

describe( 'Pruebas en RegisterScreen', () => {
    const wrapper = mount(
        <Provider store={ store }>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );

    beforeEach( () => {
        store.clearActions();
    } );

    test( 'Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    } );

    test( 'Debe ejecutar la accion setError() con el mensaje de nombre invalido', () => {
        wrapper.find( 'form' ).simulate( 'submit', {
            preventDefault(){}
        } );

        const actions = store.getActions();

        expect( actions[ 0 ] ).toEqual( {
            type: types.uiSetError,
            payload: 'Ingrese un nombre válido'
        } );
    } );

    test( 'Debe mostrar el mensaje de error con el texto correspondiente', () => {
        const store = mockStore( {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Ingrese un nombre válido'
            }
        } );

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect( wrapper.find( '.auth__alert-error' ).exists() ).toBe( true );
        expect( wrapper.find( '.auth__alert-error' ).text() ).toBe( store.getState().ui.msgError );
    } );
} );
