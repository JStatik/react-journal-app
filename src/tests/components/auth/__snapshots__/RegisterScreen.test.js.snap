// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RegisterScreen Debe mostrar el componente correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Nombre"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Contraseña"
      type="password"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Repetir contraseña"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="Submit"
    >
      Registrarse
    </button>
    <div
      className="text-center"
    >
      <a
        className="link"
        href="/auth/login"
        onClick={[Function]}
      >
        Already registered?
      </a>
    </div>
  </form>
</div>
`;
